### PR TITLE
Kubectl Data Driven Command KEP

### PR DESCRIPTION
**Pull request by pwittrock**

### CODE REVIEW

**Comment by juanvallejo** [comment]
cc @smarterclayton

**Comment by vboctor** [comment]
Looks interesting.  Few questions:

1. What would be the approach to handle the case where there is a Kubectl client version skew (Kubectl is older than cluster) doesn't understand the lasted version of the data (due to new constructs that were added to the Yaml schema).  Do you expect the API to command discovery/mounting to fail or do you expect the client to be served the older version of the command?

2. How simple are these dynamic commands going to be?  For example, do you expect to have one command == 1 call - or do you want to support multiple calls with flow control etc? The latter is much more complex to realize.

3. What would be the semantics for success, retry-later, permanent failure that the client can assume?  I assume that we would need a contract that API must comply with for the client to know what is going on.  That could be helpful for automation in a way that is agnostic for each specific command's decisions.

4. As CRDs mount commands via data, should there be a limitation that a common mounted by a specific CRD can only call the corresponding CRD APIs?  That may be better for security.

5. What would be the process for SIG CLI to stay in the loop as new commands are added via an API change to drive consistency of such dynamic commands.

6. Does that approach break or help with ability for shells to do auto-complete for kubectl?  I'm not familiar with how these work under the hood, but thought it is worth mentioning.

7. How do you see this approach surfacing in client-go?

**Comment by pwittrock** [comment]
> What would be the approach to handle the case where there is a Kubectl client version skew (Kubectl is older than cluster) doesn't understand the lasted version of the data (due to new constructs that were added to the Yaml schema). Do you expect the API to command discovery/mounting to fail or do you expect the client to be served the older version of the command?

Good question.  Ideally I envision moving toward and evergreen client with much stronger backwards compatibility than what we have today, which would address this.  In lieu of that we will need to rollout API changes as unavailable in the server, but available in the client for a full release so that -1 version skewed clients are capable of consuming the commands.

> How simple are these dynamic commands going to be? For example, do you expect to have one command == 1 call - or do you want to support multiple calls with flow control etc? The latter is much more complex to realize.

TBD.  For alpha we are just targeting single requests for create sub commands.  We can evaluate more complex workflows based on our experience with the sub commands.

> What would be the semantics for success, retry-later, permanent failure that the client can assume? I assume that we would need a contract that API must comply (or a mapping from API responses to such semantics) with for the client to know what is going on. That could be helpful for automation in a way that is agnostic for each specific command's decisions.

TBD.  For create sub commands we shouldn't attempt to retry.  I could imagine retries might be desirable for patch requests with optimistic locking failures.  Retries could configurable by the user and added in Beta.

> As CRDs mount commands via data, should there be a limitation that a command mounted by a specific CRD can only call the corresponding CRD APIs? That may be better for security.

That sounds reasonable for alpha.  For commands published by Core Resources this shouldn't be a concern.  We could also explicitly list which APIs the command makes requests to in the command description as well as support `dryrun` so users can see what requests would be made.

> What would be the process for SIG CLI to stay in the loop as new commands are added via an API change to drive consistency of such dynamic commands.

We will need to come up with a style guide and possibly a linter.

> Does that approach break or help with ability for shells to do auto-complete for kubectl? I'm not familiar with how these work under the hood, but thought it is worth mentioning.

I am not sure.  Our support for auto-complete isn't particularly strong as is.  We should figure out how much of a concern this is.

> How do you see this approach surfacing in client-go?

At this time I believe the 2 are unrelated.

**Comment by jdumars** [comment]
/hold
holding for stakeholder discussion

**Comment by pwittrock** [comment]
@vboctor Thanks for your thoughtful feedback

**Comment by justaugustus** [comment]
REMINDER: KEPs are moving to [k/enhancements](https://github.com/kubernetes/enhancements) on November 30. Please attempt to merge this KEP before then to signal consensus.
For more details on this change, review [this thread](https://groups.google.com/d/msg/kubernetes-dev/LIkZoIqCT20/t515CV2mAgAJ).

Any questions regarding this move should be directed to that thread and not asked on GitHub.

File: keps/sig-cli/0031-datadrivencommands.md [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+
+- Address cli support for extension APIs
+- Address user experience for version skewed clients
+
+### Non-Goals
+
+Allow client commands that have complex client-side logic to be data driven.
+
+- Require a TTY
+- Are Agnostic to Specific Resources
```
Comment (smarterclayton) [inline keps/sig-cli/0031-datadrivencommands.md]: Which ironically was the entire point of subresources and which has mostly failed.

File: keps/sig-cli/0031-datadrivencommands.md [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+type ResourceCommand struct {
+	// Command is the cli Command
+	Command Command `json:"command"`
+
+	// Requests are the requests the command will send to the apiserver.
+	// +optional
+	Requests []ResourceRequest `json:"requests,omitempty"`
+
+	// Output is a go-template used write the command output.  It may reference values specified as flags using
+	// {{index .Flags.Strings "flag-name"}}, {{index .Flags.Ints "flag-name"}}, {{index .Flags.Bools "flag-name"}},
```
Comment (smarterclayton) [inline keps/sig-cli/0031-datadrivencommands.md]: If this is intended to be server side only clarify.  A discussion of security risks to server side template expansion is appropriate

File: keps/sig-cli/0031-datadrivencommands.md [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+	// +optional
+	Output string `json:"output,omitempty"`
+}
+
+type ResourceOperation string
+
+const (
+	CREATE_RESOURCE ResourceOperation = "Create"
+	UPDATE_RESOURCE                   = "Update"
+	DELETE_RESOURCE                   = "Delete"
```
Comment (smarterclayton) [inline keps/sig-cli/0031-datadrivencommands.md]: Boo, hiss.  Use CamelCase, we’re not C savages here.

File: keps/sig-cli/0031-datadrivencommands.md [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+	//		- name: responsename
+	//        jsonPath: "{.metadata.name}"
+	//
+	// +optional
+	SaveResponseValues []ResponseValue `json:"saveResponseValues,omitempty"`
+}
+
+// Flag defines a cli flag that should be registered and available in request / output templates
+type Flag struct {
+	Type FlagType `json:"type"`
```
Comment (smarterclayton) [inline keps/sig-cli/0031-datadrivencommands.md]: We should align this closer to the structure of Table and be as consistent as possible.  Would prefer at minimum defining Format.

File: keps/sig-cli/0031-datadrivencommands.md [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+	//    		type: String
+	//    		stringValue: "default"
+	//    		description: "deployment namespace"
+	//
+	// +optional
+	Flags []Flag `json:"flags,omitempty"`
+
+	// SuggestFor is an array of command names for which this command will be suggested -
+	// similar to aliases but only suggests.
+	SuggestFor []string `json:"suggestFor,omitempty"`
```
Comment (smarterclayton) [inline keps/sig-cli/0031-datadrivencommands.md]: While I know why this looks like the cobra struct, something about bothers me.

File: keps/sig-cli/0031-datadrivencommands.md [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+    saveResponseValues:
+    - name: responsename
+      jsonPath: "{.metadata.name}"
+  output: |
+    deployment.apps/{{index .Responses.Strings "responsename"}} created
+```
+
+### Risks and Mitigations
+
+- Command name collisions: CRD publishes command that conflicts with another command
```
Comment (smarterclayton) [inline keps/sig-cli/0031-datadrivencommands.md]: Should follow the same rules we use for CRDs and api discovery priority.

File: keps/sig-cli/0031-datadrivencommands.md (hunk 1 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: I like the idea of subresources too...

File: keps/sig-cli/0031-datadrivencommands.md (hunk 3 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: lolz, will update

File: keps/sig-cli/0031-datadrivencommands.md (hunk 6 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: +1

File: keps/sig-cli/0031-datadrivencommands.md (hunk 4 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: Do you have a link to the format you are referring to?

File: keps/sig-cli/0031-datadrivencommands.md (hunk 2 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: I am actually proposing that the template is expanded on the client from the response.

Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: Will clarify

File: keps/sig-cli/0031-datadrivencommands.md (hunk 4 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: To clarify, this struct is for populating the client side command flags (e.g. Cobra).

File: keps/sig-cli/0031-datadrivencommands.md (hunk 5 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: We can iterate on the precise API format and capabilities.

File: keps/sig-cli/0031-datadrivencommands.md (hunk 1 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: FWIW, this approach could complement subresources that argument compiled into the client using the same approach to invoke them as it does the Resource APIs.

**Comment by pwittrock** [comment]
PTAL

File: keps/sig-cli/0031-datadrivencommands.md (hunk 6 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: Added a comment about this

File: keps/sig-cli/0031-datadrivencommands.md (hunk 4 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: Added comments about this to clarify.

File: keps/sig-cli/0031-datadrivencommands.md (hunk 3 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: Done

File: keps/sig-cli/0031-datadrivencommands.md (hunk 2 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: Done

**Comment by pwittrock** [comment]
@soltysh Did you want to weigh in on this?

File: keps/sig-cli/0031-datadrivencommands.md (hunk 4 above)
Comment (smarterclayton) [reply keps/sig-cli/0031-datadrivencommands.md]: Yeah, I meant that between flags, openapi, and the table object that server side get returns we should be consistent in how we describe and use constants.  Don’t have to be identical, but don’t drift for the sake of drift.

File: keps/sig-cli/0031-datadrivencommands.md (hunk 2 above)
Comment (smarterclayton) [reply keps/sig-cli/0031-datadrivencommands.md]: Ugh, client side expansion will only work in Go.  Are there use cases we are eliminating from by making this a client problem?  Use from web consoles?

File: keps/sig-cli/0031-datadrivencommands.md (hunk 1 above)
Comment (smarterclayton) [reply keps/sig-cli/0031-datadrivencommands.md]: Yes

File: keps/sig-cli/0031-datadrivencommands.md (hunk 5 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: One of the main goals of the PoC was to demonstrate that Cobra commands could be represented as declarative data published by the cluster.  Having proven that, re-thinking the representation and capabilities sounds like a good plan.

**Comment by pwittrock** [comment]
@kubernetes/sig-cli-api-reviews @kubernetes/sig-cli-maintainers @kubernetes/sig-cli-feature-requests 

If there are no more comments on this PR by December 5, I'd like to merge it.

File: keps/sig-cli/0031-datadrivencommands.md (hunk 2 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: Yes.  I considered those, however that is more ambitious in scope than what is being proposed here. If this were to be done server-side, I would imagine it as driven through subresources.  However, I'd like to see data to justify this as necessary given it is a significant increase in scope and complexity.

**Comment by pwittrock** [comment]
@smarterclayton Sorry for not being more responsive.  For some reason didn't see your comments.

File: keps/sig-cli/0031-datadrivencommands.md (hunk 4 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: SG.  Will add a TODO for this.

**Comment by pwittrock** [comment]
@smarterclayton Are you ok with merging before the migration of KEPs?

File: keps/sig-cli/0031-datadrivencommands.md (hunk 2 above)
Comment (pwittrock) [reply keps/sig-cli/0031-datadrivencommands.md]: I can add an alternatives considered section.

**Comment by pwittrock** [comment]
TODO: Rename the file to match the KEP number

**Pull request merged**